### DISCUSSION TITLE
Nefarius Bluetooth PS Enumerator device - Code 31

### DISCUSSION THREAD

**Discussion by niallr** [nefarius/BthPS3#25] [Support]
I have installed BthPS3 v2.1.153 and had difficulties where my Bluetooth device (listed as Generic Bluetooth Radio) would have a Code 10 error with an exclamation mark. I have fixed this as per [this link](https://forums.vigem.org/topic/408/bthps3-bluetooth-drivers_v1-2-2-installation-failure/4?_=1628097977550&lang=en-GB) that recommended deleting the "LowerFilters" item at "HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Class{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}" in the registry.

I am now facing a new issue where the Nefarius Bluetooth PS Enumerator device has an exclamation mark in Device Manager with the below error description:
> This device is not working properly because Windows cannot load the drivers required for this device. (Code 31)
> 
> Object Name not found.

I am on Windows 10 Version 10.0.19043 Build 19043

**Comment by nefarius** [comment]
Uninstall and install again, that error means the filter driver isn't installed properly for whatever reason.

**Reply by niallr** [reply]
Thanks for the prompt response. I've removed the drivers via Driver Store Explorer, uninstalled and reinstalled but still face the same issue. I've checked the "DriverStore\FileRepository" folder and can see that the 3 driver directories have been created successfully with all of the files that are included in the installer.

I've taken a screenshot of the registry entries for the Nefarius Bluetooth PS Enumerator device in case this is useful:
![image](https://user-images.githubusercontent.com/915985/128241758-9343e672-77bb-4f2f-b4ad-840929b341c2.png)

**Reply by nefarius** [reply]
Use the setup for that, it takes care of the filter driver installation. What you showed is not the filter driver, that is the profile driver.

Uninstall using the setup, reboot, install, reboot, check again.

**Reply by niallr** [reply]
I followed your recommendation and uninstalled via setup, restarted, ran the install (via admin cmd just in case), restarted, and I am receiving the Code 10 error with an exclamation mark that I mentioned in the first post.
![image](https://user-images.githubusercontent.com/915985/128258517-6171678b-0fbc-48db-a151-8277349448ef.png)
![image](https://user-images.githubusercontent.com/915985/128258553-fb042297-b732-4015-8506-f1f3e83e35dd.png)

I then removed the "LowerFilters" item like before and restarted, which then gives this in Device Manager (with the Code 31):
![image](https://user-images.githubusercontent.com/915985/128258967-19db040f-70d5-4949-b97d-3a22ba71a6e3.png)

Could this be a permissions issue within Windows preventing the installer from completing the filter driver setup?

**Reply by nefarius** [reply]
We need a [trace log](https://vigem.org/projects/BthPS3/Debugging-the-drivers/) to see what's going on.

**Reply by niallr** [reply]
Thanks, I ran a trace as per the instructions but it seems to be blank - see attached.
[BthPS3.zip](https://github.com/ViGEm/BthPS3/files/6938462/BthPS3.zip)

I then tried uninstalling and reinstalling with logging via PowerShell in case there is something useful there.
[bthps3_install-filtered.log](https://github.com/ViGEm/BthPS3/files/6938459/bthps3_install-filtered.log)

Just as the installation completes and prompts for restart, I see 2 events in Windows Event Viewer appeared:

> Event 7040: The start type of the BthPS3 Service service was changed from demand start to disabled.
> Event 7040: The start type of the Nefarius Bluetooth PS Filter Service service was changed from system start to disabled.

Once the PC is restarted, those services don't appear at all in the Services list.

**Reply by nefarius** [reply]
Yeah, it's empty. This is how you are supposed to do it:

- Install everything
- Reboot at least once
- Unplug or disable your Bluetooth host device
- Start the trace
- Plug in or enable Bluetooth host radio
- Stop the trace
- Share it

**Reply by niallr** [reply]
Thanks @nefarius , the log looks more promising now that I followed the steps in your comment - see attached.
[BthPS3.zip](https://github.com/ViGEm/BthPS3/files/6942007/BthPS3.zip)

**Reply by nefarius** [reply]
The Trace reveals that the filter driver is still not installed and working properly, therefore nothing will work:

```text
 [DriverEntry] --> Entry
 [DriverEntry] <-- Exit <status=STATUS_SUCCESS (0x00000000)>
 [BthPS3EvtDeviceAdd] --> Entry
 [BthPS3_CreateDevice] --> Entry
 [BthPS3_ServerContextInit] --> Entry
 [BthPS3_DeviceContextHeaderInit] --> Entry
 [BthPS3_DeviceContextHeaderInit] <-- Exit <status=STATUS_SUCCESS (0x00000000)>
 [BthPS3_ServerContextInit] <-- Exit <status=STATUS_SUCCESS (0x00000000)>
 [BthPS3_OpenFilterIoTarget] --> Entry
 [BthPS3_OpenFilterIoTarget] ERROR:WdfIoTargetOpen failed with status STATUS_OBJECT_NAME_NOT_FOUND (0xC0000034)
 [BthPS3_OpenFilterIoTarget] <-- Exit <status=STATUS_OBJECT_NAME_NOT_FOUND (0xC0000034)>
 [BthPS3_CreateDevice] <-- Exit <status=STATUS_OBJECT_NAME_NOT_FOUND (0xC0000034)>
 [BthPS3EvtDeviceAdd] <-- Exit <status=STATUS_OBJECT_NAME_NOT_FOUND (0xC0000034)>
 [BthPS3EvtDriverContextCleanup] --> Entry
```

Share the file `C:\Windows\INF\setupapi.dev.log`, that should hold the info what's going wrong.

**Reply by niallr** [reply]
I appreciate your persistence and support with this!

Attached the log file as requested.
[setupapi.dev.log](https://github.com/ViGEm/BthPS3/files/6945459/setupapi.dev.log)

**Reply by nefarius** [reply]
What the heck, this looks completely clean and fine too...

Hm, open Device Manager, enable `View / Show hidden devices` and open the `System devices` node, do you find this there?

![image](https://user-images.githubusercontent.com/286631/128516528-a7f87b30-1c84-4d52-b6ae-2418fe29a6ae.png)

**Comment by niallr** [comment]
Here's System Devices with hidden device shown - looks a bit odd!
![image](https://user-images.githubusercontent.com/915985/128530293-ae167aab-5920-4088-ab51-a3cf75845cae.png)


...and here is what Bluetooth looks like too:
![image](https://user-images.githubusercontent.com/915985/128530356-17fcf651-9425-488e-871f-a0f3cd3a4c98.png)

**Reply by nefarius** [reply]
Ah I see, that's not supposed to be like that. Right-click, Uninstall all those Filter entries, run the normal uninstaller, reboot, install fresh via setup again.

**Reply by niallr** [reply]
Thanks, I uninstalled all of the Filter Driver entries as well as the duplicates of the other Bluetooth items so that there were only 1 of each. I then uninstalled, restarted, installed fresh, restarted again and still seem to have the Code 34 issue. See below for a screenshot of the log for convenience:
![image](https://user-images.githubusercontent.com/915985/128630929-c7554da1-c234-4210-b50d-cbb3bd21078f.png)
[BthPS3 (2).zip](https://github.com/ViGEm/BthPS3/files/6950644/BthPS3.2.zip)

Device Manager:
![image](https://user-images.githubusercontent.com/915985/128630944-7aa4376e-96b9-48af-bbe3-bee8832e3d83.png)

![image](https://user-images.githubusercontent.com/915985/128630958-105dfe19-7a1a-4e43-b883-6d841e587ec7.png)

**Reply by nefarius** [reply]
It can't communicate with the filter driver for whatever reason. 

Check the registry location `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Class\{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}` and validate that the value `LowerFilters` with content `BthPS3PSM` is there.

Also, share the Hardware ID of your Bluetooth host radio.

**Reply by niallr** [reply]
With the LowerFilters value set to "BthPS3PSM", I get the original Code 10 error again, and the other Bluetooth devices are then hidden from the list:

![image](https://user-images.githubusercontent.com/915985/128635175-863b5e65-e9ad-44a9-a44f-3dd47b1242f4.png)

![image](https://user-images.githubusercontent.com/915985/128635147-bda7d3f9-8e12-49eb-b988-3aff8458c8d6.png)


Here is another trace while reconnecting the Bluetooth adapter (screenshot for convenience):
![image](https://user-images.githubusercontent.com/915985/128635266-04278b39-3dec-4023-b73b-167ced3d033a.png)
[BthPS3 (3).zip](https://github.com/ViGEm/BthPS3/files/6950939/BthPS3.3.zip)



Device details:
![image](https://user-images.githubusercontent.com/915985/128634978-df52226f-5809-4275-b1ac-53714e30605f.png)
Device USB\VID_0A12&PID_0001\6&3af0f9ce&0&13

I had purchased the Bluetooth adapter from a different marketplace, but [this](https://www.ebay.co.uk/itm/USB-Bluetooth-V5-0-Wireless-Mini-Dongle-Adapter-For-Win7-8-10-PC-Laptop/123011436969) is the same product. I am using the driver that Windows automatically identified for it and didn't install any third-party drivers.

**Reply by nefarius** [reply]
That's bad, the filter driver can't boot, for whatever reason and without the whole solution can't function.

**Reply by niallr** [reply]
Hmm, very strange. I ran `sfc /scannow` in case of any system file issues that might be causing it, but no change there.

Just to check, do the reg entries below look correct? This is just after a complete re-install but without removing the LowerFilters value (so basically the same situation as [mentioned here with the Code 10 error on the Bluetooth radio](https://github.com/ViGEm/BthPS3/discussions/25#discussioncomment-1144635).
![image](https://user-images.githubusercontent.com/915985/128640343-3cd7fbea-15ea-4512-8a2a-489fac2ca6bc.png)

![image](https://user-images.githubusercontent.com/915985/128640357-50d208de-0910-4bcd-8bd1-4bdcb218951c.png)

![image](https://user-images.githubusercontent.com/915985/128640366-74e9dd67-3eb1-4111-ad02-c4fef1f38c89.png)

![image](https://user-images.githubusercontent.com/915985/128640481-09220673-bef3-4f10-a091-38aa286680fc.png)

**Reply by nefarius** [reply]
LGTM

**Reply by niallr** [reply]
Thanks. I'm going to try another Bluetooth adapter to try and isolate whether this is an issue specific to my OS install or the adapter. Will keep you posted. Thank you for all the help so far!

**Reply by TranceInvasion** [reply]
Hello Everyone :D

Im posting cause im suffering the same problem with my bth usb dongle with the same HW ID. 
I did all the mencioned solution here, but still cant make it work.
![image](https://user-images.githubusercontent.com/101219491/157342057-32064480-39e5-4b3e-801b-d1b476ee6c55.png)

This its an image of my dongle:

<img src="https://user-images.githubusercontent.com/101219491/157342997-e414d30a-0e02-4163-9954-1c748cb959e9.png" width="100" height="130">

**Reply by nefarius** [reply]
IIRC the issue with this dongle couldn't be resolved.

**Comment by oJeff95** [comment]
Hello, I used the BthPS3 normally with a bluetooth 4.0 dongle of the same ID normally, but it is not very good, I can only use a maximum of 3 devices at the same time, and it also has many flaws with little distance without barriers.
So I decided to buy a new standard 5.1 bluetooth dongle, model Baseus BA04, which supports up to 7 devices at the same time and has the same ID too, I connected to the PC and had the same error as above. I figured I would have to uninstall everything with the old bluetooth and install it again with the new one, I did that but I had the same problem.

I tried removing all related hidden drivers and doing a clean install again and had the same issue.
So I did a fresh clean install with bluetooth 4.0, connected my controller via cable and then via bluetooth, then unplugged the controller and dongle. So I found this thread and did the filter removal with powershell code. I connected bluetooth 5.1 and there was no problem starting the drivers.

But the PS3 controller couldn't connect via bluetooth, so by chance I opened DS4Windows and realized that the ViGEM drivers had been deleted for some reason too, I reinstalled them, disconnected and connected the bluetooth dongle and I got error code 31 on the Nefarios Bluetooth PS Enumerator device.

So I removed the bluetooth 5.1 and connected the bluetooth 4.0 again to check and I was having the same problem.
After re-entering the filter with this command:

New-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\Class\{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}' -Name 'LowerFilters' -PropertyType MultiString -Value BthPS3PSM

And remove and replace bluetooth 4.0, the error disappears, but bluetooth 5.1 does not start, it has the same error as always, Generic Bluetooth Radio device did not start correctly, code 10.

Anyway, Bluetooth 5.1 can only start if the filter is removed, but both my dongles can only start correctly without the filter if the ViGEM drivers are uninstalled.

With this report, do you think there is a solution? I stayed up all night trying to make it work anyway =X.

Thanks for the great work on everything!
(Translated by google translator from Portuguese-Brazil)

New dongle link: https://pt.aliexpress.com/item/1005003809655059.html
Old dongle link (very similar to this one, but 4.0): https://pt.aliexpress.com/item/1005003853724980.html?spm=a2g0o.productlist.0.0.29f2301fYlgTlm&algo_pvid=ece6bc7e-0ffd-4452-a43a-8c6a1cae0512&algo_exp_id=ece6bc7e-0ffd-4452-a43a-8c6a1cae0512-11&pdp_ext_f=%7B%22sku_id%22%3A%2212000027340489467%22%7D&pdp_pi=-1%3B27.9%3B-1%3B-1%40salePrice%3BBRL%3Bsearch-mainSearch

**Comment by Milor123** [comment]
and what is the solution guys? I have the same problem T_T and my device is compatible. is a "Generic RADIO"

**Reply by nefarius** [reply]
[Read the FAQ, it's covered there](https://vigem.org/projects/BthPS3/Frequently-Asked-Questions/#how-to-fix-bluetooth-device-error-codes-19-39).

**Reply by oJeff95** [reply]
I couldn't solve.  Currently I use a Bluetooth 4.0 dongle to use the DS3, and when I want to use more controls, I have to disable the filter and use my bluetooth 5.1 dongle.
Note: Both dongles are detected as "generic radio"
Note²: I haven't tested the new version yet.  As soon as I'm home, I'll test it out again.

**Reply by Milor123** [reply]
@nefarius  I've tried it but not works, the bluethooth now working but not Nefarius, is bugged. I try reinstall it like 4 try
@JeffS2Duda I have the same problem that you, the mine is 5.0 dongle

![image](https://user-images.githubusercontent.com/14153649/190400544-419fc206-0fee-4750-a440-cfab9534e0a3.png)

Hardware id:
```
USB\VID_0A12&PID_0001&REV_8891
USB\VID_0A12&PID_0001
```

Is it a windows problem? what should i do?
My windows is 21H2 19044.1319 Windows 10

and when try open configuration tool:
![image](https://user-images.githubusercontent.com/14153649/190401848-00b5171d-2a40-487f-bc01-d46f76c1fa1c.png)

Ammm a detail that you should be know... For that windows detect my bluethooth and works as normal, I always must run this line in powershell after of install Bthps3

`Remove-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\Class\{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}' -Name 'LowerFilters'
`

**Note**: Add a image of my register
![image](https://user-images.githubusercontent.com/14153649/190405159-4432b94d-b0ee-46f8-9631-05b40fef0a3a.png)

**Note 2**: When I disconnect my bluetooth of my USB Port, Now i can open BTS3 tool
Only open with bluethooth disconnected from ports:
![image](https://user-images.githubusercontent.com/14153649/190406260-1137fe25-d0c7-4dff-a666-99e1414d83f4.png)

**VirtualMachine test 1** : A TEST IN A NEW VIRTUAL MACHINE (CLEAN INSTALLATION)

Again i need run `Remove-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\Class\{e0cbf06c-cd8b-4647-bb8a-263b43f0f974}' -Name 'LowerFilters'
` or the bluethooth keep stuck

Now, the difference in this installation is that **Nefarius Bluetooth PS Enumerator** not appear in the bluethooth list
check this:
![image](https://user-images.githubusercontent.com/14153649/190418762-31ff9bb7-045d-4072-a505-5f233f7c0212.png)


**VirtualMachine Note**: When uninstall and reinstall the **Nefarius Bluetooth PS Enumerator** is now visible but bugged, and after of reboot keep with the same problem.
![image](https://user-images.githubusercontent.com/14153649/190419933-ba471d19-4a3b-4792-be78-1c75e4836020.png)

**Reply by nefarius** [reply]
No idea, works on thousands of installations.

**Reply by Milor123** [reply]
> No idea, works on thousands of installations.

Uwu Surely is the device, Anything for depurate it bro? what do you suggest me?

**Reply by nefarius** [reply]
Everything I have to say is on the site already documented. I have nothing new to add.

**Reply by nefarius** [reply]
You can ask on our Discord if somebody had a similar issue, I don't have that kind of time to look into every individual case.

**Reply by Milor123** [reply]
> Everything I have to say is on the site already documented. I have nothing new to add.

Ohh understand, thank you very much. The solution for this problem, simply I can said, Is **change the device** in my case i am sure because, I've gifted my old bluethooth to a friend, now he plays with BTSPS3, but now I want play and i've get other device of other brand, and there are the problem. Try search for other brand and test.

**Reply by Milor123** [reply]
> You can ask on our Discord if somebody had a similar issue, I don't have that kind of time to look into every individual case.

Please share us your discord team please !

**Reply by nefarius** [reply]
It is so easy to find and linked everywhere important I won't repeat it here. See it as a little exercise 😉

**Reply by oJeff95** [reply]
I updated VigemBus and BthPS3 to the latest version, my 4.0 dongle kept working normally and my 5.1 dongle now gives me code 10.
When I remove the filter, I can at least use the 5.1 dongle with other things, but the "Nefarius Bluetooth PS Enumerator" shows the code 31 =/.

Dongle 5.1
![image](https://user-images.githubusercontent.com/64566344/190449653-27e2baa2-7399-4b29-aed1-8eb829e11a6c.png)

Apparently the only solution is to use another dongle.
So, I don't recommend you to buy the Baseus Bluetooth Dongle ba04 5.1 if you want to use BthPS3. (Although I vaguely remember seeing in some video someone using version 5.0 of this dongle working normally.)

Anyway, thanks for the help!

Dongle 4.0
![image](https://user-images.githubusercontent.com/64566344/190449379-57dd3773-06d4-4c8b-948f-3921c82afb26.png)

**Reply by r4f4xD** [reply]
link del discord ?

**Reply by Milor123** [reply]
https://discord.vigem.org/

**Reply by nefarius** [reply]
Party-pooper, spoon-feeding doesn't help people grow their skills 😛
